File: dimos/memory/codecs/cdr.py
# Copyright 2026 Dimensional Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from dimos.msgs.protocol import DimosMsg


class CdrCodec:
    """Store a generated message as its transport-neutral CDR bytes."""

    def __init__(self, msg_type: type[DimosMsg]) -> None:
        if not all(
            hasattr(msg_type, member) for member in ("encode", "decode", "msg_name", "schema")
        ):
            raise TypeError(f"{msg_type!r} is not a generated CDR message type")
        self._msg_type = msg_type

    @property
    def payload_type(self) -> type[DimosMsg]:
        """Message type encoded by this codec."""
        return self._msg_type

    def encode(self, value: DimosMsg) -> bytes:
        if not isinstance(value, self._msg_type):
            raise TypeError(f"Expected {self._msg_type.msg_name}, got {type(value).__name__}")
        return value.encode()

    def decode(self, data: bytes) -> DimosMsg:
        return self._msg_type.decode(data)

File: native/rust/dimos-module/src/cdr.rs
// Copyright 2026 Dimensional Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Copyright 2026 Dimensional Inc.
// SPDX-License-Identifier: Apache-2.0

//! Generated CDR codecs for native module ports.
use dimos_generated_messages::codec::Message;
use std::io;

pub fn encode<T: Message>(message: &T) -> io::Result<Vec<u8>> {
    message.encode()
}

pub fn decode<T: Message>(bytes: &[u8]) -> io::Result<T> {
    T::decode(bytes)
}

#[cfg(test)]
mod tests {
    use super::*;
    use dimos_generated_messages::{geometry_msgs::msg::PoseStamped, std_msgs::msg::Int32};

    #[test]
    fn nested_fields_and_both_byte_orders() {
        let mut message = PoseStamped::default();
        message.header.frame_id = "map".into();
        message.header.stamp.sec = 1700000000;
        message.header.stamp.nanosec = 123456789;
        message.pose.position.x = 1.5;
        message.pose.orientation.w = 1.0;
        assert_eq!(
            decode::<PoseStamped>(&encode(&message).unwrap()).unwrap(),
            message
        );
        assert_eq!(
            decode::<PoseStamped>(&message.encode_endian(false).unwrap()).unwrap(),
            message
        );
    }

    #[test]
    fn malformed_payload_is_a_decode_error() {
        let mut bytes = encode(&Int32 { data: 1234 }).unwrap();
        assert_eq!(
            decode::<Int32>(&bytes[..bytes.len() - 1])
                .unwrap_err()
                .kind(),
            io::ErrorKind::InvalidData
        );
        bytes.push(0);
        assert_eq!(
            decode::<Int32>(&bytes).unwrap_err().kind(),
            io::ErrorKind::InvalidData
        );
        bytes[0] = 255;
        assert_eq!(
            decode::<Int32>(&bytes).unwrap_err().kind(),
            io::ErrorKind::InvalidData
        );
    }
}
